### PR TITLE
Frontapp refactor and receive custom messages action

### PR DESCRIPTION
**Pull request by jcortes**

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/frontapp/actions/import-message/import-message.mjs [hunk 1]
```diff
@@ -1,16 +1,17 @@
...
-import { axios } from "@pipedream/platform";
+import frontApp from "../../frontapp.app.mjs";
+import utils from "../../common/utils.mjs";
 
 export default {
   key: "frontapp-import-message",
   name: "Import Message",
-  description: "Appends a new message into an inbox.",
-  version: "0.1.2",
+  description: "Appends a new message into an inbox. [See the docs here](https://dev.frontapp.com/reference/import-inbox-message",
```
Comment (michelle0927) [inline components/frontapp/actions/import-message/import-message.mjs:7]: The ending parenthesis is missing.
```suggestion
  description: "Appends a new message into an inbox. [See the docs here](https://dev.frontapp.com/reference/import-inbox-message)",
```

File: components/frontapp/actions/import-message/import-message.mjs [hunk 2]
```diff
@@ -1,16 +1,17 @@
...
+  version: "0.1.4",
```
Comment (michelle0927) [inline components/frontapp/actions/import-message/import-message.mjs:8]: Shouldn't this be version 0.1.3?
```suggestion
  version: "0.1.3",
```

File: components/frontapp/actions/receive-custom-messages/receive-custom-messages.mjs [hunk 3]
```diff
@@ -0,0 +1,108 @@
+import utils from "../../common/utils.mjs";
+import frontApp from "../../frontapp.app.mjs";
+
+export default {
+  key: "frontapp-send-new-message",
```
Comment (michelle0927) [inline components/frontapp/actions/receive-custom-messages/receive-custom-messages.mjs:5]:
```suggestion
  key: "frontapp-receive-custom-messages",
```

File: components/frontapp/actions/receive-custom-messages/receive-custom-messages.mjs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+  name: "Send new message",
```
Comment (michelle0927) [inline components/frontapp/actions/receive-custom-messages/receive-custom-messages.mjs:6]:
```suggestion
  name: "Receive Custom Messages",
```

File: components/frontapp/actions/receive-custom-messages/receive-custom-messages.mjs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+  description: "Sends a new message from a channel. It will create a new conversation. [See the docs here](https://dev.frontapp.com/reference/post_channels-channel-id-messages).",
+  version: "0.2.3",
```
Comment (michelle0927) [inline components/frontapp/actions/receive-custom-messages/receive-custom-messages.mjs:8]: Shouldn't the version be 0.0.1?
```suggestion
  version: "0.0.1",
```

File: components/frontapp/actions/send-new-message/send-new-message.mjs [hunk 6]
```diff
@@ -1,27 +1,24 @@
-// legacy_hash_id: a_poikPY
-import { axios } from "@pipedream/platform";
+import utils from "../../common/utils.mjs";
+import frontApp from "../../frontapp.app.mjs";
 
 export default {
   key: "frontapp-send-new-message",
   name: "Send new message",
```
Comment (michelle0927) [inline components/frontapp/actions/send-new-message/send-new-message.mjs:6]: Names should be capitalized like titles
```suggestion
  name: "Send New Message",
```

File: components/frontapp/actions/send-new-message/send-new-message.mjs [hunk 7]
```diff
@@ -1,27 +1,24 @@
...
-  description: "Sends a new message from a channel. It will create a new conversation.",
-  version: "0.2.1",
+  description: "Sends a new message from a channel. It will create a new conversation. [See the docs here](https://dev.frontapp.com/reference/post_channels-channel-id-messages).",
+  version: "0.2.3",
```
Comment (michelle0927) [inline components/frontapp/actions/send-new-message/send-new-message.mjs:8]: Shouldn't the version be 0.2.2?
```suggestion
  version: "0.2.2",
```

File: components/frontapp/actions/update-conversation/update-conversation.mjs [hunk 8]
```diff
@@ -1,23 +1,24 @@
...
+import utils from "../../common/utils.mjs";
+import frontApp from "../../frontapp.app.mjs";
 
 export default {
   key: "frontapp-update-conversation",
   name: "Update Conversation",
-  description: "Updates a conversation",
-  version: "0.1.2",
+  description: "Updates a conversation. [See the docs here](https://dev.frontapp.com/reference/patch_conversations-conversation-id)",
+  version: "0.1.4",
```
Comment (michelle0927) [inline components/frontapp/actions/update-conversation/update-conversation.mjs:8]: Shouldn't the version be 0.1.3?
```suggestion
  version: "0.1.3",
```

**Review by michelle0927 — CHANGES_REQUESTED** [review]

File: components/frontapp/actions/import-message/import-message.mjs [hunk 9]
```diff
@@ -49,20 +50,17 @@ export default {
...
+        "bodyFormat",
       ],
     },
-    external_id: {
+    externalId: {
       type: "string",
       description: "External identifier of the message. Front won't import two messages with the same external ID.",
     },
-    created_at: {
+    createdAt: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/import-message/import-message.mjs:63]: Can this be a `optional` prop with default as `new Date().getTime()`?

File: components/frontapp/actions/import-message/import-message.mjs [hunk 10]
```diff
@@ -77,94 +75,101 @@ export default {
         "custom",
       ],
     },
-    assignee_id: {
+    assigneeId: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/import-message/import-message.mjs:78]: Can this support `async options`?

File: components/frontapp/actions/import-message/import-message.mjs [hunk 11]
```diff
@@ -77,94 +75,101 @@ export default {
...
       type: "string",
       description: "ID of the teammate who will be assigned to the conversation.",
       optional: true,
     },
     tags: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/import-message/import-message.mjs:83]: Can this support `async options`?

File: components/frontapp/actions/send-new-message/send-new-message.mjs [hunk 12]
```diff
@@ -1,27 +1,24 @@
...
+  version: "0.2.2",
   type: "action",
   props: {
-    frontapp: {
-      type: "app",
-      app: "frontapp",
-    },
-    channel_id: {
+    frontApp,
+    channelId: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/send-new-message/send-new-message.mjs:12]: Can this support `async options`?

File: components/frontapp/actions/send-new-message/send-new-message.mjs [hunk 13]
```diff
@@ -1,27 +1,24 @@
...
       type: "string",
       description: "Id or address of the channel from which to send the message",
     },
-    author_id: {
+    authorId: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/send-new-message/send-new-message.mjs:16]: Can this support `async options`?

File: components/frontapp/actions/send-new-message/send-new-message.mjs [hunk 14]
```diff
@@ -41,62 +38,74 @@ export default {
...
+      ],
     },
-    options: {
-      type: "object",
-      description: "Sending options",
-      optional: true,
-    },
-    options_tags: {
-      type: "any",
+    optionsTagIds: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/send-new-message/send-new-message.mjs:46]: Can this support `async options`?

File: components/frontapp/actions/update-conversation/update-conversation.mjs [hunk 15]
```diff
@@ -1,23 +1,24 @@
...
-  version: "0.1.2",
+  description: "Updates a conversation. [See the docs here](https://dev.frontapp.com/reference/patch_conversations-conversation-id).",
+  version: "0.1.3",
   type: "action",
   props: {
-    frontapp: {
-      type: "app",
-      app: "frontapp",
+    frontApp,
+    conversationId: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/update-conversation/update-conversation.mjs:12]: Can this support `async options`?

File: components/frontapp/actions/update-conversation/update-conversation.mjs [hunk 16]
```diff
@@ -1,23 +1,24 @@
...
+      type: "string",
+      description: "Conversation unique identifier",
     },
-    assignee_id: {
+    assigneeId: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/update-conversation/update-conversation.mjs:16]: Can this support `async options` and maybe adding a `null` to the end of the list to unassign?

File: components/frontapp/actions/update-conversation/update-conversation.mjs [hunk 17]
```diff
@@ -1,23 +1,24 @@
...
       type: "string",
       description: "ID of the teammate to assign the conversation to. Set it to null to unassign.",
       optional: true,
     },
-    inbox_id: {
+    inboxId: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/update-conversation/update-conversation.mjs:21]: Can this support `async options`?

File: components/frontapp/actions/update-conversation/update-conversation.mjs [hunk 18]
```diff
@@ -33,42 +34,39 @@ export default {
         "open",
       ],
     },
-    tag_ids: {
-      type: "any",
+    tagIds: {
```
Comment (andrewjschuang) [inline components/frontapp/actions/update-conversation/update-conversation.mjs:37]: Can this support `async options`?

**Review by andrewjschuang — CHANGES_REQUESTED** [review]
@jcortes thanks for the submission!

I've left some comments below, most of them to support `async options`

File: components/frontapp/actions/import-message/import-message.mjs (hunk 9 above)
Comment (jcortes) [reply components/frontapp/actions/import-message/import-message.mjs:63]: Actually is a required field

**Review by vellames-turing — DISMISSED** [review]
lgtm

**Comment by DilanAthukorala** [comment]
@jcortes Observing the following issues;

 for **receive_custom_messages** action

1.   unable to input thread ref / values not returned

image



2. Attachments are not sent to the front app (tried different base64 format files) even though the request is successful



for **import_message** action following issue observed
image



for **send_new_message** action following issues are observed
image



for **update_conversation**  action following issues are observed 

1. following issue observed when sending request

2.  unable to set assignee value as null

image


3. Conversation id's are not retrieved (how user can find these id's) and unable to set null value for it

image

**Comment by jcortes** [comment]
Hi @DilanAthukorala can you please try to test again all components thanks!

**Comment by jcortes** [comment]
Hi @DilanAthukorala I think I've fixed all these issue now. Can you test again please 🙏

**Comment by DilanAthukorala** [comment]
Verification completed.

**Review by andrewjschuang — APPROVED** [review]

**Pull request merged**